core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/util/OrderComparator.java

```diff
@@ -45,17 +49,47 @@ public OrderComparator(EvaluationStrategy strategy, Order order, ValueComparator
 			QueryEvaluationContext context) {
 		this.cmp = vcmp;
 		Comparator<BindingSet> allComparator = null;
+
 		for (OrderElem element : order.getElements()) {
-			final QueryValueEvaluationStep prepared = strategy.precompile(element.getExpr(), context);
-			final boolean ascending = element.isAscending();
-			Comparator<BindingSet> comparator = (o1, o2) -> {
-				Value v1 = prepared.evaluate(o1);
-				Value v2 = prepared.evaluate(o2);
-
-				int compare = cmp.compare(v1, v2);
-				return ascending ? compare : -compare;
-			};
+			boolean ascending = element.isAscending();
+			ValueExpr expr = element.getExpr();
+			Comparator<BindingSet> comparator;
+
+			if (expr instanceof Var) {
+				// Here we optimize for the most common case where the ORDER BY clause uses Var(s) e.g. "ORDER BY ?a"
+
+				Function<BindingSet, Value> getValue = context.getValue(((Var) expr).getName());
+
+				comparator = (o1, o2) -> {
+					Value v1 = getValue.apply(o1);
+					Value v2 = getValue.apply(o2);
+
+					int compare = cmp.compare(v1, v2);
+					return ascending ? compare : -compare;
+				};
+			} else {
```

Comment:
@kenwenzel I made this code here to handle the more common case of simply ordering by one or more variables. Here the upstream code returns null instead of throwing an exception.